src/V3Ast.cpp

```diff
@@ -631,11 +631,53 @@ void AstNode::relinkOneLink(AstNode*& pointpr,  // Ref to pointer that gets set
 }
 
 void AstNode::addHereThisAsNext(AstNode* newp) {
-    // {old}->this->{next} becomes {old}->new->this->{next}
-    VNRelinker handle;
-    this->unlinkFrBackWithNext(&handle);
-    newp->addNext(this);
-    handle.relink(newp);
+    // {back}->this->{next} becomes {back}->new->this->{next}
+    UASSERT(!newp->backp(), "New node already linked?");
+    AstNode* const endp = newp->m_headtailp;
+    UASSERT(endp, "m_headtailp not set on new node");
+    AstNode* const backp = this->m_backp;
+    //
+    this->editCountInc();
+    // Common linkage
+    endp->m_nextp = this;
+    this->m_backp = endp;
+    newp->m_backp = backp;
+    endp->m_headtailp = nullptr;
+    // Linkage dependent on position
+    if (backp && backp->m_nextp == this) {
+        // If 'this' is not at the head of a list, then the new node will also not be at the head
+        // of a list, so we can just link in the new node in the middle.
+        backp->m_nextp = newp;
+        newp->m_headtailp = nullptr;
+    } else {
+        // If 'this' is at the head of a list, then the new node becomes the head of that list.
+        if (backp) {
+            if (backp->m_op1p == this) {
+                backp->m_op1p = newp;
+            } else if (backp->m_op2p == this) {
+                backp->m_op2p = newp;
+            } else if (backp->m_op3p == this) {
+                backp->m_op3p = newp;
+            } else {
+                UASSERT(backp->m_op4p == this, "Don't know where newp should go");
+                backp->m_op4p = newp;
+            }
+        }
+        // We also need to update m_headtailp.
+        AstNode* const tailp = this->m_headtailp;
+        this->m_headtailp = nullptr;
+        newp->m_headtailp = tailp;
+        tailp->m_headtailp = newp;
+    }
+    // Iterator fixup
+    if (endp->m_iterpp) {
+        *(endp->m_iterpp) = this;
+        return;
+    }
```

Comment:
I don't think the old code would have done this, if anything I'd think newp->m_iterpp would be the one needing checking, but likewise don't think old code did that either.  Can you explain why added it?

Reply:
I believe `newp->addNext(this);` in the old code did this: 
https://github.com/verilator/verilator/blob/63507e8e29c0c0e1f659d7f7d0cb12205feb8158/src/V3Ast.cpp#L268
I.e.: if we are iterating on the last element of the list we are inserting before this, continue with this.
I think this is not needed for correctness by the current codebase though.